apiserver/facades/client/application/application_unit_test.go

```diff
@@ -282,6 +347,43 @@ postgresql:
 	})
 }
 
+func (s *ApplicationSuite) TestLXDProfileSetCharmWithNewerAgentVersion(c *gc.C) {
+	s.SetFeatureFlags(feature.InstanceMutater)
+
+	err := s.api.SetCharm(params.ApplicationSetCharm{
+		ApplicationName: "postgresql",
+		CharmURL:        "cs:postgresql",
+		ConfigSettings:  map[string]string{"stringOption": "value"},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	s.backend.CheckCallNames(c, "Application", "Charm")
+	s.backend.charm.CheckCallNames(c, "Config")
+	app := s.backend.applications["postgresql"]
+	app.CheckCallNames(c, "AgentTools", "AllUnits", "SetCharmProfile", "SetCharm")
+	// We don't care about 0, 1 calls, as they take no arguments.
+	app.CheckCall(c, 2, "SetCharmProfile", "cs:postgresql")
+	app.CheckCall(c, 3, "SetCharm", state.SetCharmConfig{
+		Charm:          &state.Charm{},
+		ConfigSettings: charm.Settings{"stringOption": "value"},
+	})
+}
+
+func (s *ApplicationSuite) TestLXDProfileSetCharmWithOldAgentVersion(c *gc.C) {
+	s.SetFeatureFlags(feature.InstanceMutater)
+
+	err := s.api.SetCharm(params.ApplicationSetCharm{
+		ApplicationName: "redis",
```

Comment:
I'd expect the charm upgrade with redis to succeed in the real world.  Then a new test to fail with a charm having a lxd profile.